Port from speak/espeak shim to sugar3.speech

nswarup14 opened the pull request:
Fixes #8 

**Changes**
1) `createcardpanel.py`
2) `face.py`
3) `speak/face.py`
4) `speak/voice.py`
5) `speak/espeak.py` (deleted)
6) `cardlist.py`
7) `activity.py`
8) `card.py`
9) `cardtable.py`
10) `widgets.py`

**Updates**
The activity used a shim speak/espeak to make use of the gst-plugin-espeak for voice synthesis i.e from text to sound. The sugar3 toolkit has this feature in sugar3.speech, hence the patch removes the use of the shim and uses sugar3.speech instead.

**Testing**
Upon testing, the activity ran as expected and there were sounds as well. 

Thanks!

quozl commented:
Tested e8aed38 on Ubuntu 19.04.  I heard appropriate sound, as soon as I asked for it.  My test method was to click on the gear icon to _Edit game_, then click on a tile, then on the face speak toggle button.  Or right-click and use the play button in the palette.

You can test the gst-espeak-plugin package like this;
```bash
gst-launch-1.0 espeak text="hello world" ! autoaudiosink
```

If you were using virtual machine instead of native testing, the virtualisation software may be negotiating access to the host sound device.

Reviewed as at e8aed38;

* [ ] could it be that `_peding` should be `_pending`?
* [ ] did these changes arise from some other patch in another activity?  If so, credit that in commit message,
* [ ] `SpeechManager` is instantiated twice; once should be plenty, as it has many system resources at play.

nswarup14 commented:
@quozl 
Yes, I am using a VM to test my changes. The could explain the delay in the sounds.
I shall test it according to the procedure you have described as well.

>  could it be that _peding should be _pending?

I'm not sure. The variable is not used anywhere else apart from `speak/face.py`. the `update()` method also does shed light on what the variable could be. I feel we can let it be as it is.

> did these changes arise from some other patch in another activity?

No, I referred to my earlier work itself which is a closed [PR](https://github.com/sugarlabs/memorize-activity/pull/20). I did not use anyone else's work.

> `SpeechManager` is instantiated twice; once should be plenty, as it has many system resources at play

So the `SpeechManager` is instantiated in two different modules, `speak/face.py` and `speak/voice.py`. In the latter, the `SpeechManager` is instantiated inside a module function that is used elsewhere in the activity.
How can I eliminate one of the `SpeechManager` instances?

Thanks!

quozl commented:
> > could it be that _peding should be _pending?
> 
> I'm not sure. The variable is not used anywhere else apart from `speak/face.py`. the `update()` method also does shed light on what the variable could be. I feel we can let it be as it is.

Perhaps you might look at which commit it arrived in?  Use the blame view, or `git blame`, and review the commits responsible.

> > did these changes arise from some other patch in another activity?
> 
> No, I referred to my earlier work itself which is a closed [PR](https://github.com/sugarlabs/memorize-activity/pull/20). I did not use anyone else's work.

Okay, thanks.  I remember doing espeak to sugar3.speech conversions last year, and wondered how similar the code would be.

> > `SpeechManager` is instantiated twice; once should be plenty, as it has many system resources at play
> 
> So the `SpeechManager` is instantiated in two different modules, `speak/face.py` and `speak/voice.py`. In the latter, the `SpeechManager` is instantiated inside a module function that is used elsewhere in the activity.
> How can I eliminate one of the `SpeechManager` instances?

Instantiate it only in one module, then use that instance by passing it to the other modules somehow.  A suitable module might be the Activity class initialiser.

nswarup14 commented:
@quozl 
The commit that introduces `self._peding` is https://github.com/sugarlabs/memorize-activity/commit/57c2df2c5617d8f6c65e66ad2980cda681181011. However, the commit does not shed light on whether the variable was named correctly or if it was supposed to be `self._pending`.

I have removed the two instances of `SpeechManager` and replaced it with just one, that is instantiated at the beginning of the activity and passed along to the modules that require it.

I was able to reproduce the sounds instantly this time, following the procedure outlined in [comment](https://github.com/sugarlabs/memorize-activity/pull/25#issuecomment-509112678). I edited my initial comment accordingly.

Thanks

quozl commented:
Thanks.  It was from Speak activity.  https://github.com/sugarlabs/speak/commit/7a27c93ba07e707d21a8d35f73aa9733fb51eda8  From that context, it was a timing issue, and so "pending" seems likely.  I think Aleksey did not have English as first language, and was mostly active between 2009 and 2012, with a brief resurgence in 2017.

quozl commented:
https://github.com/sugarlabs/memorize-activity/pull/25/commits/5f54462bfdbd0598288fc5c6d69a672b9281b9dd is an invasive change, and adds complexity.  I suggest you revert it, and instead make a speechmanager.py which holds a global variable which is returned by a get_speechmanager function.  We've several examples of that design pattern in Sugar.

nswarup14 commented:
@quozl @chimosky
I have included sugar3.speech as discussed above in the comments.
Please review.

quozl commented:
Thanks.

Some review comments added to https://github.com/quozl/memorize-activity/commits/pr-25

I've not finished review, but that's how far I got in time available.

Please also rebase your pull request branch against _master_ branch.

nswarup14 commented:
@quozl 
I have rebased against _master_
I reviewed you the commits you added to your fork and subsequently pull them to my branch.

I have added some more commits. Tested on Ubuntu 18.04 and Sugar v0.114.
No errors in the logs and the activity runs alright.

Thanks

chimosky commented:
Tested. Works as expected, still reviewing though.

quozl commented:
Thanks.  Tested 586b1d3.

It seems the purpose of right-click menu on face icon is to select the voice language, in turn so that memorisation games can be created for bilingual learning.  Code already in voice.py is very complex, and I'm not sure why.

quozl commented on speak/face.py:
```diff
@@ -152,20 +144,37 @@ def update(self, status=None):
+            voice_name = self.status.voice.name
+        else:
+            voice_name = self._pending.voice.name
+        all_voices = self.speech.get_all_voices()
```
Several lines of `say` look just like several lines of `say_notification`.

nswarup14 replied:
I have resolved that by creating a separate function thats called from both `say` and `say_notification`.

quozl commented on speak/face.py:
```diff
@@ -152,20 +144,37 @@ def update(self, status=None):
 
     def say_notification(self, something):
-        status = (self._peding or self.status).clone()
+        status = (self._pending or self.status).clone()
```
A clone of a Status object is created, then discarded by the end of this function.  I don't see why the attributes of the original Status object are obtained instead of using a clone.  Any idea why?

nswarup14 replied:
@quozl 
I'm not sure why.
`say_notification` is used to select the language from the combobox in the game settings, whereas `say` is used when the card flip is made.

In the `say_notification` function, the `voice` attribute of the status object is being reassigned. Perhaps the idea was to prevent sideeffects being introduced by using the same `Status` object for both the above mentioned tasks.

quozl replied:
I'll work backwards and explain what I see as of e71e3be42dec84409af16f8b3318809598aeb19e.

`say_text` is to be given something to say, and a language code for the voice to use to say it.  Rather than store the language code, the activity stores the voice name.  Converting the voice name to the language code uses a `for` loop.

`say` is to be given something to say, and it will use the currently chosen voice name for the card; which it obtains from an instance of `Status`.

`say_notification` is to be given something to say, and it will use the `defaultVoice()`, which it obtains from a cloned instance of `Status`.

Just speculating here, but perhaps `say_notification` could call `speech.say_text` without a language code, causing the default voice or system language to be used.

quozl commented on speak/voice.py:
```diff
@@ -110,7 +111,24 @@ def __cmp__(self, other):
+    out = []  # Voices that produce sound only
+
+    for lang_code, name in all_voices.items():
+        if(name in ('en-rhotic', 'english_rp', 'english_wmids')):
```
Unnecessary parentheses.

nswarup14 replied:
I have removed them in the next patch.
Thanks

quozl commented on speak/voice.py:
```diff
@@ -110,7 +111,24 @@ def __cmp__(self, other):
+
+    for lang_code, name in all_voices.items():
+        if(name in ('en-rhotic', 'english_rp', 'english_wmids')):
+            # these noices do not produce sound
```
Typo, noices should be voices?  Do they really not produce sound?

nswarup14 replied:
Yes, indeed.
I have changed it.

nswarup14 replied:
I haven't been able to check whether the voices really do not produce sound. I shall try to test it

quozl replied:
Thanks.  You added this code in the commit, so I thought you already knew.  If you got it from somewhere else, tell me?

quozl commented on speak/voice.py:
```diff
@@ -110,7 +111,24 @@ def __cmp__(self, other):
+            # these noices do not produce sound
+            continue
+
+        if('_' in lang_code):
```
Unnecessary parentheses.

nswarup14 replied:
I have removed it in the next patch.

nswarup14 commented:
@quozl 
> Code already in voice.py is very complex, and I'm not sure why.
I agree. 
I cant think of a way to simplify it. It uses regex to compare languages, which a complex task.

pro-panda commented:
> I cant think of a way to simplify it. It uses regex to compare languages, which a complex task.

If the code cannot be simplified, and if you are interested, then add docstrings and comments to explain the flow

quozl commented:
The complex code in `voice.py` predates the simplification done for adding _sugar3.speech_.  I agree with @pro-panda, a good way to prove if code is needed is to document it fully, as this informs you what it does in great detail.  I've often written wonderful documentation which has let me discard both the code and documentation and replace it with something much simpler.  Remember the value of the work is in the understanding and simplifying change without removing functionality ... and not in the preservation and enlargement of unnecessary code.

quozl commented:
(I've not yet understood the code in `voice.py`, and we have an outstanding request for help from Paraguay, see [mailing list post](http://lists.sugarlabs.org/archive/sugar-devel/2019-July/057076.html).)

quozl commented on speak/face.py:
```diff
@@ -152,20 +144,35 @@ def update(self, status=None):
+        all_voices = self.speech.get_all_voices()
+        self.say_text(something, voice_name, all_voices)
+
+    def say_text(self, something, voice_name, all_voices):
```
Because I pointed out the similarity of the code leading up to two calls to say_text.  This is a variant of say_text that uses a voice name not a language code.  To avoid the interpretation that this is a redefinition, it could be named say_text_by_voice_name, or my_say_text, or just _say.  say_notification could change to say_text_in_default_language; as it took me a while to figure out why it was called notification.

nswarup14 commented:
@quozl @chimosky 
I am trying to understand the code in `voice.py` and following @pro-panda 's suggestion, I shall add some documentation based on what I understand.
I wanted to know how I can change my language from English to Spanish, so that I can reproduce the error as mentioned in the [mailing list post](http://lists.sugarlabs.org/archive/sugar-devel/2019-July/057076.html)
I shall test with both _master_ as well as with https://github.com/sugarlabs/memorize-activity/pull/25/commits/e71e3be42dec84409af16f8b3318809598aeb19e

Thanks.

quozl commented:
Before starting the activity, set the LANG environment variable to
es_PY.utf, see issue #26 title.

quozl commented:
Ah, my mistake.  I had reproduced the error reported by Aldo by installing a fresh Ubuntu virtual machine in Spanish language mode.  Something during the installation sequence _removes_ support for other languages, and I think this may be the cause of the error.

nswarup14 commented:
@quozl 
Does this mean the issue is not with the activity but rather in the installation process?

quozl commented:
No, it means it is an issue with multiple causes, and we must then choose which cause to fix and justify it.  Do you really think that when Ubuntu is installed in Spanish that it should not delete English files?  I think that would be difficult to justify.

nswarup14 commented:
@quozl @chimosky 
How do I proceed with the PR and ensure it's fixed correctly?
Thanks

quozl commented:
We need a developer who can speak Spanish and English, and neither of us is that developer, so we'll leave that problem to them for now.
